{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "85aea5b9-37d7-4e78-ab78-72f0cee3b6c2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2025-08-15 10:01:44.025882] TEgenomeSimulator started.\n",
      "[2025-08-15 10:01:44.025962] Arguments: {'mode': 0, 'to_mask': False, 'prefix': 'reference_simulation1', 'repeat': 'data/fake_repeat.fa', 'repeat2': None, 'maxcp': 10, 'mincp': 1, 'maxidn': 95, 'minidn': 80, 'maxsd': 20, 'minsd': 1, 'chridx': 'data/random_genome_chr_index_small.csv', 'genome': None, 'alpha': 0.5, 'beta': 0.7, 'intact': 0.001, 'seed': 1, 'threads': 1, 'outdir': 'data'}\n",
      "Output Directory: /mnt/c/Users/james/projects/calisim-examples/workshop/tegenomesimulator/data/TEgenomeSimulator_reference_simulation1_result\n",
      "Mode: 0\n",
      "Running Random Synthesized mode.\n",
      "Prefix: reference_simulation1\n",
      "Repeat: data/fake_repeat.fa\n",
      "Chromosome Index: data/random_genome_chr_index_small.csv\n",
      "Genome File: None\n",
      "Alpha: 0.5\n",
      "Beta: 0.7\n",
      "Max Copies: 10\n",
      "Min Copies: 1\n",
      "Upper bound of mean identity: 95\n",
      "Lower bound of mean identity: 80\n",
      "Upper bound of sd of mean identity: 20\n",
      "Lower bound of sd of mean ideneity: 1\n",
      "Max chance of intact insertion: 0.001\n",
      "Seed: 1\n",
      "\n",
      "\n",
      "#########################################################\n",
      "### Prepare TE library table with simulation settings ###\n",
      "#########################################################\n",
      "Using repeat fasta file data/fake_repeat.fa\n",
      "Output directory set as /mnt/c/Users/james/projects/calisim-examples/workshop/tegenomesimulator/data/TEgenomeSimulator_reference_simulation1_result\n",
      "\n",
      "\n",
      "## Random chosing copy numbe for each TE family ##\n",
      "Maximum copy number set by user: 10\n",
      "Minimum copy number set by user: 10\n",
      "\n",
      "\n",
      "## Random chosing the averaged sequence identity for each TE family ##\n",
      "Maximum averaged sequence identity: 95\n",
      "Minimum averaged sequence identity: 80\n",
      "\n",
      "\n",
      "## Random chosing the standard deviation of averaged sequence identity for each TE family ##\n",
      "Maximum standard deviation of averaged sequence identity: 20\n",
      "Minimum standard deviation of averaged sequence identity: 1\n",
      "\n",
      "\n",
      "## Random chosing the proportion of INDEL to total SNP (dependant on sequence identity) for each TE family ##\n",
      "Maximum INDEL proportion set by default: 20\n",
      "Minimum INDEL proportion set by default: 5\n",
      "\n",
      "\n",
      "## Setting the length of std based on prior knowledge ##\n",
      "Length range of TSD for LTR retrotransposon set to: 5 - 5\n",
      "Length range of TSD for LINE set to: 5 - 20\n",
      "Length range of TSD for SINE set to: 5 - 20\n",
      "Length range of TSD for DTA set to: 5 - 8\n",
      "Length range of TSD for DTC set to: 2 - 4\n",
      "Length range of TSD for DTH set to: 3 - 3\n",
      "Length range of TSD for DTM set to: 8 - 9\n",
      "Length range of TSD for DTT set to: 2 - 2\n",
      "Length range of TSD for Helitron set to: 0\n",
      "Length range of TSD for MITE set to: 2 - 10\n",
      "Length range of TSD for else set to: 0\n",
      "\n",
      "\n",
      "## Extracting the length of each TE family ##\n",
      "\n",
      "\n",
      "## Setting the proportion of fragmented TE loci of each TE family ##\n",
      "Maximum chance of keeping a TE insertion intact as 100% integrity in each TE family: 0.001\n",
      "Minimum chance of keeping a TE insertion intact in each TE family: 0\n",
      "Maximum proportion of fragmented TE loci of each TE family: 100\n",
      "Minimum proportion of fragmented TE loci of each TE family: 99.999\n",
      "\n",
      "\n",
      "## Setting the proportion of nested TE insertion of each Copia or Gypsy family ##\n",
      "Maximum proportion of nested TE insertion of each Copia or Gypsy family set by default: 30\n",
      "Minimum proportion of nested TE insertion of each Copia or Gypsy family set by default: 0\n",
      "\n",
      "\n",
      "## Printing the TE library table ##\n",
      "Generated the TE library table for simulation. File saved as /mnt/c/Users/james/projects/calisim-examples/workshop/tegenomesimulator/data/TEgenomeSimulator_reference_simulation1_result/TElib_sim_list.table\n",
      "\n",
      "\n",
      "\n",
      "TE library table generated successfully. Output logged to /mnt/c/Users/james/projects/calisim-examples/workshop/tegenomesimulator/data/TEgenomeSimulator_reference_simulation1_result/TEgenomeSimulator.log\n",
      "mode=0, running prep_yml_config.py for Random Genome Mode.\n",
      "\n",
      "\n",
      "#############################################\n",
      "### Prepare TEgenomeSimulator config file ###\n",
      "#############################################\n",
      "Using genome fasta file None\n",
      "Using repeat fasta file data/fake_repeat.fa\n",
      "Output directory set as /mnt/c/Users/james/projects/calisim-examples/workshop/tegenomesimulator/data/TEgenomeSimulator_reference_simulation1_result\n",
      "Generated the config file for simulation. File saved as /mnt/c/Users/james/projects/calisim-examples/workshop/tegenomesimulator/data/TEgenomeSimulator_reference_simulation1_result/TEgenomeSimulator_reference_simulation1.yml\n",
      "\n",
      "\n",
      "\n",
      "Config file generated successfully. Output logged to /mnt/c/Users/james/projects/calisim-examples/workshop/tegenomesimulator/data/TEgenomeSimulator_reference_simulation1_result/TEgenomeSimulator.log\n",
      "\n",
      "\n",
      "##############################################################\n",
      "### Mutate TE sequence and perform non-overlap TE insertion###\n",
      "##############################################################\n",
      "Using mode 0 (0 for random genome; 1 for custome genome)\n",
      "Using config file TEgenomeSimulator_reference_simulation1.yml\n",
      "\n",
      "Genome with non-overlap random TE insertions was generated successfully. Output logged to /mnt/c/Users/james/projects/calisim-examples/workshop/tegenomesimulator/data/TEgenomeSimulator_reference_simulation1_result/TEgenomeSimulator.log\n",
      "\n",
      "\n",
      "##############################################################\n",
      "### Mutate TE sequence and perform non-overlap TE insertion###\n",
      "##############################################################\n",
      "Using mode 0 (0 for random genome; 1 for custome genome)\n",
      "Using config file TEgenomeSimulator_reference_simulation1.yml\n",
      "/home/jbris/.cache/pypoetry/virtualenvs/calisim-examples-yfUJKvMn-py3.10/lib/python3.10/site-packages/Bio/SeqRecord.py:354: BiopythonDeprecationWarning: Using a string as the sequence is deprecated and will raise a TypeError in future. It has been converted to a Seq object.\n",
      "  warnings.warn(\n",
      "\n",
      "Genome with non-overlap random and nested TE insertions was generated successfully. Output logged to /mnt/c/Users/james/projects/calisim-examples/workshop/tegenomesimulator/data/TEgenomeSimulator_reference_simulation1_result/TEgenomeSimulator.log\n",
      "[2025-08-15 10:01:46.757121] TEgenomeSimulator started.\n",
      "[2025-08-15 10:01:46.757327] Arguments: {'mode': 0, 'to_mask': False, 'prefix': 'reference_simulation2', 'repeat': 'data/fake_repeat.fa', 'repeat2': None, 'maxcp': 10, 'mincp': 1, 'maxidn': 95, 'minidn': 80, 'maxsd': 20, 'minsd': 1, 'chridx': 'data/random_genome_chr_index_small.csv', 'genome': None, 'alpha': 0.5, 'beta': 0.7, 'intact': 0.001, 'seed': 1, 'threads': 1, 'outdir': 'data'}\n",
      "Output Directory: /mnt/c/Users/james/projects/calisim-examples/workshop/tegenomesimulator/data/TEgenomeSimulator_reference_simulation2_result\n",
      "Mode: 0\n",
      "Running Random Synthesized mode.\n",
      "Prefix: reference_simulation2\n",
      "Repeat: data/fake_repeat.fa\n",
      "Chromosome Index: data/random_genome_chr_index_small.csv\n",
      "Genome File: None\n",
      "Alpha: 0.5\n",
      "Beta: 0.7\n",
      "Max Copies: 10\n",
      "Min Copies: 1\n",
      "Upper bound of mean identity: 95\n",
      "Lower bound of mean identity: 80\n",
      "Upper bound of sd of mean identity: 20\n",
      "Lower bound of sd of mean ideneity: 1\n",
      "Max chance of intact insertion: 0.001\n",
      "Seed: 1\n",
      "\n",
      "\n",
      "#########################################################\n",
      "### Prepare TE library table with simulation settings ###\n",
      "#########################################################\n",
      "Using repeat fasta file data/fake_repeat.fa\n",
      "Output directory set as /mnt/c/Users/james/projects/calisim-examples/workshop/tegenomesimulator/data/TEgenomeSimulator_reference_simulation2_result\n",
      "\n",
      "\n",
      "## Random chosing copy numbe for each TE family ##\n",
      "Maximum copy number set by user: 10\n",
      "Minimum copy number set by user: 10\n",
      "\n",
      "\n",
      "## Random chosing the averaged sequence identity for each TE family ##\n",
      "Maximum averaged sequence identity: 95\n",
      "Minimum averaged sequence identity: 80\n",
      "\n",
      "\n",
      "## Random chosing the standard deviation of averaged sequence identity for each TE family ##\n",
      "Maximum standard deviation of averaged sequence identity: 20\n",
      "Minimum standard deviation of averaged sequence identity: 1\n",
      "\n",
      "\n",
      "## Random chosing the proportion of INDEL to total SNP (dependant on sequence identity) for each TE family ##\n",
      "Maximum INDEL proportion set by default: 20\n",
      "Minimum INDEL proportion set by default: 5\n",
      "\n",
      "\n",
      "## Setting the length of std based on prior knowledge ##\n",
      "Length range of TSD for LTR retrotransposon set to: 5 - 5\n",
      "Length range of TSD for LINE set to: 5 - 20\n",
      "Length range of TSD for SINE set to: 5 - 20\n",
      "Length range of TSD for DTA set to: 5 - 8\n",
      "Length range of TSD for DTC set to: 2 - 4\n",
      "Length range of TSD for DTH set to: 3 - 3\n",
      "Length range of TSD for DTM set to: 8 - 9\n",
      "Length range of TSD for DTT set to: 2 - 2\n",
      "Length range of TSD for Helitron set to: 0\n",
      "Length range of TSD for MITE set to: 2 - 10\n",
      "Length range of TSD for else set to: 0\n",
      "\n",
      "\n",
      "## Extracting the length of each TE family ##\n",
      "\n",
      "\n",
      "## Setting the proportion of fragmented TE loci of each TE family ##\n",
      "Maximum chance of keeping a TE insertion intact as 100% integrity in each TE family: 0.001\n",
      "Minimum chance of keeping a TE insertion intact in each TE family: 0\n",
      "Maximum proportion of fragmented TE loci of each TE family: 100\n",
      "Minimum proportion of fragmented TE loci of each TE family: 99.999\n",
      "\n",
      "\n",
      "## Setting the proportion of nested TE insertion of each Copia or Gypsy family ##\n",
      "Maximum proportion of nested TE insertion of each Copia or Gypsy family set by default: 30\n",
      "Minimum proportion of nested TE insertion of each Copia or Gypsy family set by default: 0\n",
      "\n",
      "\n",
      "## Printing the TE library table ##\n",
      "Generated the TE library table for simulation. File saved as /mnt/c/Users/james/projects/calisim-examples/workshop/tegenomesimulator/data/TEgenomeSimulator_reference_simulation2_result/TElib_sim_list.table\n",
      "\n",
      "\n",
      "\n",
      "TE library table generated successfully. Output logged to /mnt/c/Users/james/projects/calisim-examples/workshop/tegenomesimulator/data/TEgenomeSimulator_reference_simulation2_result/TEgenomeSimulator.log\n",
      "mode=0, running prep_yml_config.py for Random Genome Mode.\n",
      "\n",
      "\n",
      "#############################################\n",
      "### Prepare TEgenomeSimulator config file ###\n",
      "#############################################\n",
      "Using genome fasta file None\n",
      "Using repeat fasta file data/fake_repeat.fa\n",
      "Output directory set as /mnt/c/Users/james/projects/calisim-examples/workshop/tegenomesimulator/data/TEgenomeSimulator_reference_simulation2_result\n",
      "Generated the config file for simulation. File saved as /mnt/c/Users/james/projects/calisim-examples/workshop/tegenomesimulator/data/TEgenomeSimulator_reference_simulation2_result/TEgenomeSimulator_reference_simulation2.yml\n",
      "\n",
      "\n",
      "\n",
      "Config file generated successfully. Output logged to /mnt/c/Users/james/projects/calisim-examples/workshop/tegenomesimulator/data/TEgenomeSimulator_reference_simulation2_result/TEgenomeSimulator.log\n",
      "\n",
      "\n",
      "##############################################################\n",
      "### Mutate TE sequence and perform non-overlap TE insertion###\n",
      "##############################################################\n",
      "Using mode 0 (0 for random genome; 1 for custome genome)\n",
      "Using config file TEgenomeSimulator_reference_simulation2.yml\n",
      "\n",
      "Genome with non-overlap random TE insertions was generated successfully. Output logged to /mnt/c/Users/james/projects/calisim-examples/workshop/tegenomesimulator/data/TEgenomeSimulator_reference_simulation2_result/TEgenomeSimulator.log\n",
      "\n",
      "\n",
      "##############################################################\n",
      "### Mutate TE sequence and perform non-overlap TE insertion###\n",
      "##############################################################\n",
      "Using mode 0 (0 for random genome; 1 for custome genome)\n",
      "Using config file TEgenomeSimulator_reference_simulation2.yml\n",
      "/home/jbris/.cache/pypoetry/virtualenvs/calisim-examples-yfUJKvMn-py3.10/lib/python3.10/site-packages/Bio/SeqRecord.py:354: BiopythonDeprecationWarning: Using a string as the sequence is deprecated and will raise a TypeError in future. It has been converted to a Seq object.\n",
      "  warnings.warn(\n",
      "\n",
      "Genome with non-overlap random and nested TE insertions was generated successfully. Output logged to /mnt/c/Users/james/projects/calisim-examples/workshop/tegenomesimulator/data/TEgenomeSimulator_reference_simulation2_result/TEgenomeSimulator.log\n"
     ]
    }
   ],
   "source": [
    "! python /mnt/c/Users/james/projects/TEgenomeSimulator/TEgenomeSimulator/TEgenomeSimulator.py -M 0 -p reference_simulation1 -r data/fake_repeat.fa -c data/random_genome_chr_index_small.csv -o data\n",
    "! python /mnt/c/Users/james/projects/TEgenomeSimulator/TEgenomeSimulator/TEgenomeSimulator.py -M 0 -p reference_simulation2 -r data/fake_repeat.fa -c data/random_genome_chr_index_small.csv -o data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "584edc69-3858-4ff8-8f8f-6f5d8018242e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2025-08-15 10:50:42.376574] TEgenomeSimulator started.\n",
      "[2025-08-15 10:50:42.376671] Arguments: {'mode': 0, 'to_mask': False, 'prefix': 'reference_simulation1', 'repeat': 'data/fake_repeat.fa', 'repeat2': None, 'maxcp': 10, 'mincp': 1, 'maxidn': 95, 'minidn': 80, 'maxsd': 20, 'minsd': 1, 'chridx': 'data/random_genome_chr_index_small.csv', 'genome': None, 'alpha': 0.5, 'beta': 0.7, 'intact': 0.001, 'seed': 1, 'threads': 1, 'outdir': 'data'}\n",
      "Output Directory: /mnt/c/Users/james/projects/calisim-examples/workshop/tegenomesimulator/data/TEgenomeSimulator_reference_simulation1_result\n",
      "Mode: 0\n",
      "Running Random Synthesized mode.\n",
      "Prefix: reference_simulation1\n",
      "Repeat: data/fake_repeat.fa\n",
      "Chromosome Index: data/random_genome_chr_index_small.csv\n",
      "Genome File: None\n",
      "Alpha: 0.5\n",
      "Beta: 0.7\n",
      "Max Copies: 10\n",
      "Min Copies: 1\n",
      "Upper bound of mean identity: 95\n",
      "Lower bound of mean identity: 80\n",
      "Upper bound of sd of mean identity: 20\n",
      "Lower bound of sd of mean ideneity: 1\n",
      "Max chance of intact insertion: 0.001\n",
      "Seed: 1\n",
      "\n",
      "\n",
      "#########################################################\n",
      "### Prepare TE library table with simulation settings ###\n",
      "#########################################################\n",
      "Using repeat fasta file data/fake_repeat.fa\n",
      "Output directory set as /mnt/c/Users/james/projects/calisim-examples/workshop/tegenomesimulator/data/TEgenomeSimulator_reference_simulation1_result\n",
      "\n",
      "\n",
      "## Random chosing copy numbe for each TE family ##\n",
      "Maximum copy number set by user: 10\n",
      "Minimum copy number set by user: 10\n",
      "\n",
      "\n",
      "## Random chosing the averaged sequence identity for each TE family ##\n",
      "Maximum averaged sequence identity: 95\n",
      "Minimum averaged sequence identity: 80\n",
      "\n",
      "\n",
      "## Random chosing the standard deviation of averaged sequence identity for each TE family ##\n",
      "Maximum standard deviation of averaged sequence identity: 20\n",
      "Minimum standard deviation of averaged sequence identity: 1\n",
      "\n",
      "\n",
      "## Random chosing the proportion of INDEL to total SNP (dependant on sequence identity) for each TE family ##\n",
      "Maximum INDEL proportion set by default: 20\n",
      "Minimum INDEL proportion set by default: 5\n",
      "\n",
      "\n",
      "## Setting the length of std based on prior knowledge ##\n",
      "Length range of TSD for LTR retrotransposon set to: 5 - 5\n",
      "Length range of TSD for LINE set to: 5 - 20\n",
      "Length range of TSD for SINE set to: 5 - 20\n",
      "Length range of TSD for DTA set to: 5 - 8\n",
      "Length range of TSD for DTC set to: 2 - 4\n",
      "Length range of TSD for DTH set to: 3 - 3\n",
      "Length range of TSD for DTM set to: 8 - 9\n",
      "Length range of TSD for DTT set to: 2 - 2\n",
      "Length range of TSD for Helitron set to: 0\n",
      "Length range of TSD for MITE set to: 2 - 10\n",
      "Length range of TSD for else set to: 0\n",
      "\n",
      "\n",
      "## Extracting the length of each TE family ##\n",
      "\n",
      "\n",
      "## Setting the proportion of fragmented TE loci of each TE family ##\n",
      "Maximum chance of keeping a TE insertion intact as 100% integrity in each TE family: 0.001\n",
      "Minimum chance of keeping a TE insertion intact in each TE family: 0\n",
      "Maximum proportion of fragmented TE loci of each TE family: 100\n",
      "Minimum proportion of fragmented TE loci of each TE family: 99.999\n",
      "\n",
      "\n",
      "## Setting the proportion of nested TE insertion of each Copia or Gypsy family ##\n",
      "Maximum proportion of nested TE insertion of each Copia or Gypsy family set by default: 30\n",
      "Minimum proportion of nested TE insertion of each Copia or Gypsy family set by default: 0\n",
      "\n",
      "\n",
      "## Printing the TE library table ##\n",
      "Generated the TE library table for simulation. File saved as /mnt/c/Users/james/projects/calisim-examples/workshop/tegenomesimulator/data/TEgenomeSimulator_reference_simulation1_result/TElib_sim_list.table\n",
      "\n",
      "\n",
      "\n",
      "TE library table generated successfully. Output logged to /mnt/c/Users/james/projects/calisim-examples/workshop/tegenomesimulator/data/TEgenomeSimulator_reference_simulation1_result/TEgenomeSimulator.log\n",
      "mode=0, running prep_yml_config.py for Random Genome Mode.\n",
      "\n",
      "\n",
      "#############################################\n",
      "### Prepare TEgenomeSimulator config file ###\n",
      "#############################################\n",
      "Using genome fasta file None\n",
      "Using repeat fasta file data/fake_repeat.fa\n",
      "Output directory set as /mnt/c/Users/james/projects/calisim-examples/workshop/tegenomesimulator/data/TEgenomeSimulator_reference_simulation1_result\n",
      "Generated the config file for simulation. File saved as /mnt/c/Users/james/projects/calisim-examples/workshop/tegenomesimulator/data/TEgenomeSimulator_reference_simulation1_result/TEgenomeSimulator_reference_simulation1.yml\n",
      "\n",
      "\n",
      "\n",
      "Config file generated successfully. Output logged to /mnt/c/Users/james/projects/calisim-examples/workshop/tegenomesimulator/data/TEgenomeSimulator_reference_simulation1_result/TEgenomeSimulator.log\n",
      "\n",
      "\n",
      "##############################################################\n",
      "### Mutate TE sequence and perform non-overlap TE insertion###\n",
      "##############################################################\n",
      "Using mode 0 (0 for random genome; 1 for custome genome)\n",
      "Using config file TEgenomeSimulator_reference_simulation1.yml\n",
      "\n",
      "Genome with non-overlap random TE insertions was generated successfully. Output logged to /mnt/c/Users/james/projects/calisim-examples/workshop/tegenomesimulator/data/TEgenomeSimulator_reference_simulation1_result/TEgenomeSimulator.log\n",
      "\n",
      "\n",
      "##############################################################\n",
      "### Mutate TE sequence and perform non-overlap TE insertion###\n",
      "##############################################################\n",
      "Using mode 0 (0 for random genome; 1 for custome genome)\n",
      "Using config file TEgenomeSimulator_reference_simulation1.yml\n",
      "/home/jbris/.cache/pypoetry/virtualenvs/calisim-examples-yfUJKvMn-py3.10/lib/python3.10/site-packages/Bio/SeqRecord.py:354: BiopythonDeprecationWarning: Using a string as the sequence is deprecated and will raise a TypeError in future. It has been converted to a Seq object.\n",
      "  warnings.warn(\n",
      "\n",
      "Genome with non-overlap random and nested TE insertions was generated successfully. Output logged to /mnt/c/Users/james/projects/calisim-examples/workshop/tegenomesimulator/data/TEgenomeSimulator_reference_simulation1_result/TEgenomeSimulator.log\n",
      "[2025-08-15 10:50:44.820598] TEgenomeSimulator started.\n",
      "[2025-08-15 10:50:44.820670] Arguments: {'mode': 0, 'to_mask': False, 'prefix': 'reference_simulation2', 'repeat': 'data/fake_repeat.fa', 'repeat2': None, 'maxcp': 10, 'mincp': 1, 'maxidn': 95, 'minidn': 80, 'maxsd': 20, 'minsd': 1, 'chridx': 'data/random_genome_chr_index_small.csv', 'genome': None, 'alpha': 0.5, 'beta': 0.7, 'intact': 0.001, 'seed': 1, 'threads': 1, 'outdir': 'data'}\n",
      "Output Directory: /mnt/c/Users/james/projects/calisim-examples/workshop/tegenomesimulator/data/TEgenomeSimulator_reference_simulation2_result\n",
      "Mode: 0\n",
      "Running Random Synthesized mode.\n",
      "Prefix: reference_simulation2\n",
      "Repeat: data/fake_repeat.fa\n",
      "Chromosome Index: data/random_genome_chr_index_small.csv\n",
      "Genome File: None\n",
      "Alpha: 0.5\n",
      "Beta: 0.7\n",
      "Max Copies: 10\n",
      "Min Copies: 1\n",
      "Upper bound of mean identity: 95\n",
      "Lower bound of mean identity: 80\n",
      "Upper bound of sd of mean identity: 20\n",
      "Lower bound of sd of mean ideneity: 1\n",
      "Max chance of intact insertion: 0.001\n",
      "Seed: 1\n",
      "\n",
      "\n",
      "#########################################################\n",
      "### Prepare TE library table with simulation settings ###\n",
      "#########################################################\n",
      "Using repeat fasta file data/fake_repeat.fa\n",
      "Output directory set as /mnt/c/Users/james/projects/calisim-examples/workshop/tegenomesimulator/data/TEgenomeSimulator_reference_simulation2_result\n",
      "\n",
      "\n",
      "## Random chosing copy numbe for each TE family ##\n",
      "Maximum copy number set by user: 10\n",
      "Minimum copy number set by user: 10\n",
      "\n",
      "\n",
      "## Random chosing the averaged sequence identity for each TE family ##\n",
      "Maximum averaged sequence identity: 95\n",
      "Minimum averaged sequence identity: 80\n",
      "\n",
      "\n",
      "## Random chosing the standard deviation of averaged sequence identity for each TE family ##\n",
      "Maximum standard deviation of averaged sequence identity: 20\n",
      "Minimum standard deviation of averaged sequence identity: 1\n",
      "\n",
      "\n",
      "## Random chosing the proportion of INDEL to total SNP (dependant on sequence identity) for each TE family ##\n",
      "Maximum INDEL proportion set by default: 20\n",
      "Minimum INDEL proportion set by default: 5\n",
      "\n",
      "\n",
      "## Setting the length of std based on prior knowledge ##\n",
      "Length range of TSD for LTR retrotransposon set to: 5 - 5\n",
      "Length range of TSD for LINE set to: 5 - 20\n",
      "Length range of TSD for SINE set to: 5 - 20\n",
      "Length range of TSD for DTA set to: 5 - 8\n",
      "Length range of TSD for DTC set to: 2 - 4\n",
      "Length range of TSD for DTH set to: 3 - 3\n",
      "Length range of TSD for DTM set to: 8 - 9\n",
      "Length range of TSD for DTT set to: 2 - 2\n",
      "Length range of TSD for Helitron set to: 0\n",
      "Length range of TSD for MITE set to: 2 - 10\n",
      "Length range of TSD for else set to: 0\n",
      "\n",
      "\n",
      "## Extracting the length of each TE family ##\n",
      "\n",
      "\n",
      "## Setting the proportion of fragmented TE loci of each TE family ##\n",
      "Maximum chance of keeping a TE insertion intact as 100% integrity in each TE family: 0.001\n",
      "Minimum chance of keeping a TE insertion intact in each TE family: 0\n",
      "Maximum proportion of fragmented TE loci of each TE family: 100\n",
      "Minimum proportion of fragmented TE loci of each TE family: 99.999\n",
      "\n",
      "\n",
      "## Setting the proportion of nested TE insertion of each Copia or Gypsy family ##\n",
      "Maximum proportion of nested TE insertion of each Copia or Gypsy family set by default: 30\n",
      "Minimum proportion of nested TE insertion of each Copia or Gypsy family set by default: 0\n",
      "\n",
      "\n",
      "## Printing the TE library table ##\n",
      "Generated the TE library table for simulation. File saved as /mnt/c/Users/james/projects/calisim-examples/workshop/tegenomesimulator/data/TEgenomeSimulator_reference_simulation2_result/TElib_sim_list.table\n",
      "\n",
      "\n",
      "\n",
      "TE library table generated successfully. Output logged to /mnt/c/Users/james/projects/calisim-examples/workshop/tegenomesimulator/data/TEgenomeSimulator_reference_simulation2_result/TEgenomeSimulator.log\n",
      "mode=0, running prep_yml_config.py for Random Genome Mode.\n",
      "\n",
      "\n",
      "#############################################\n",
      "### Prepare TEgenomeSimulator config file ###\n",
      "#############################################\n",
      "Using genome fasta file None\n",
      "Using repeat fasta file data/fake_repeat.fa\n",
      "Output directory set as /mnt/c/Users/james/projects/calisim-examples/workshop/tegenomesimulator/data/TEgenomeSimulator_reference_simulation2_result\n",
      "Generated the config file for simulation. File saved as /mnt/c/Users/james/projects/calisim-examples/workshop/tegenomesimulator/data/TEgenomeSimulator_reference_simulation2_result/TEgenomeSimulator_reference_simulation2.yml\n",
      "\n",
      "\n",
      "\n",
      "Config file generated successfully. Output logged to /mnt/c/Users/james/projects/calisim-examples/workshop/tegenomesimulator/data/TEgenomeSimulator_reference_simulation2_result/TEgenomeSimulator.log\n",
      "\n",
      "\n",
      "##############################################################\n",
      "### Mutate TE sequence and perform non-overlap TE insertion###\n",
      "##############################################################\n",
      "Using mode 0 (0 for random genome; 1 for custome genome)\n",
      "Using config file TEgenomeSimulator_reference_simulation2.yml\n",
      "\n",
      "Genome with non-overlap random TE insertions was generated successfully. Output logged to /mnt/c/Users/james/projects/calisim-examples/workshop/tegenomesimulator/data/TEgenomeSimulator_reference_simulation2_result/TEgenomeSimulator.log\n",
      "\n",
      "\n",
      "##############################################################\n",
      "### Mutate TE sequence and perform non-overlap TE insertion###\n",
      "##############################################################\n",
      "Using mode 0 (0 for random genome; 1 for custome genome)\n",
      "Using config file TEgenomeSimulator_reference_simulation2.yml\n",
      "/home/jbris/.cache/pypoetry/virtualenvs/calisim-examples-yfUJKvMn-py3.10/lib/python3.10/site-packages/Bio/SeqRecord.py:354: BiopythonDeprecationWarning: Using a string as the sequence is deprecated and will raise a TypeError in future. It has been converted to a Seq object.\n",
      "  warnings.warn(\n",
      "\n",
      "Genome with non-overlap random and nested TE insertions was generated successfully. Output logged to /mnt/c/Users/james/projects/calisim-examples/workshop/tegenomesimulator/data/TEgenomeSimulator_reference_simulation2_result/TEgenomeSimulator.log\n"
     ]
    }
   ],
   "source": [
    "! tegenomesimulator -M 0 -p reference_simulation1 -r data/fake_repeat.fa -c data/random_genome_chr_index_small.csv -o data\n",
    "! tegenomesimulator -M 0 -p reference_simulation2 -r data/fake_repeat.fa -c data/random_genome_chr_index_small.csv -o data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ee90fd3d-0515-4dfc-9c06-ab9ca53663be",
   "metadata": {},
   "outputs": [],
   "source": [
    "! cmp data/TEgenomeSimulator_reference_simulation1_result/reference_simulation1_genome_sequence_out_final.fasta data/TEgenomeSimulator_reference_simulation2_result/reference_simulation2_genome_sequence_out_final.fasta "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d27782c0-e9d1-49f1-953c-05a03785ecf6",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
